kernel/aster-nix/src/vm/vmo/mod.rs

```diff
@@ -127,219 +135,370 @@ bitflags! {
+/// 1. File-backed Vmo: the Vmo backed by a file and resides in the `PageCache`,
+/// which includes a pager to provide it with actual pages.
+/// 2. Anonymous Vmo: the Vmo without a file backup, which does not have a pager.
 pub(super) struct Vmo_ {
+    pager: Option<Arc<dyn Pager>>,
     /// Flags
     flags: VmoFlags,
-    /// VmoInner
-    inner: Mutex<VmoInner>,
+    /// The offset of the range of pages corresponding to the vmo within `pages`.
+    page_idx_offset: usize,
+    /// The contiguous pages where the Vmo resides. Pages are managed by XArray,
+    /// and can be shared among slice children.
+    pages: Arc<Mutex<XArray<VmFrame, VmoMark>>>,
+    /// The size of current `Vmo_`.
+    size: AtomicUsize,
 }
 
-struct VmoInner {
-    pager: Option<Arc<dyn Pager>>,
-    /// size, in bytes
-    size: usize,
-    /// The pages committed. The key is the page index, the value is the committed frame.
-    committed_pages: BTreeMap<usize, VmFrame>,
-    /// The pages from the parent that current vmo can access. The pages can only be inherited when create childs vmo.
-    /// We store the page index range
-    inherited_pages: Option<Vec<VmFrame>>,
-    /// Whether the vmo is copy on write child.
-    is_cow: bool,
+fn clone_page(page: &VmFrame) -> Result<VmFrame> {
+    let new_page = VmAllocOptions::new(1).alloc_single()?;
+    new_page.copy_from_frame(page);
+    Ok(new_page)
 }
 
-impl VmoInner {
-    fn commit_page(&mut self, offset: usize) -> Result<()> {
-        let page_idx = offset / PAGE_SIZE;
-        // Fast path: the page is already committed.
-        if self.committed_pages.contains_key(&page_idx) {
-            return Ok(());
-        }
-        let frame = match &self.pager {
-            None => VmAllocOptions::new(1).alloc_single()?,
-            Some(pager) => pager.commit_page(page_idx)?,
+impl Vmo_ {
+    /// Prepare a new `VmFrame` for the target index in pages, returning the new page as well as
+    /// whether this page needs to be marked as exclusive.
+    ///
+    /// Based on the type of `Vmo` and the impending operation on the prepared page, there are 3 conditions:
+    /// 1. For an Anonymous `Vmo`, provide a new page directly. If the Vmo requires copy-on-write (COW),
+    ///    the prepared page can be directly set to exclusive.
+    /// 2. For a File-backed `Vmo` that does not need to trigger the COW mechanism,
+    ///    obtain a page from the pager directly without the need to be set as exclusive.
+    /// 3. For a File-backed `Vmo` that requires triggering the COW mechanism, obtain a page
+    ///    from the pager and then copy it. This page can be set as exclusive.
+    fn prepare_page(
+        &self,
+        page_idx: usize,
+        is_cow_vmo: bool,
+        will_write: bool,
+    ) -> Result<(VmFrame, bool)> {
+        let (page, should_mark_exclusive) = match &self.pager {
+            None => {
+                // Condition 1.
+                (VmAllocOptions::new(1).alloc_single()?, is_cow_vmo)
```

Comment:
https://github.com/asterinas/asterinas/pull/632#discussion_r1517213154
>  Otherwise, the new page does not need to set this mark since this mark is only meaningful for pages in cow vmo.

OK, but I still feel that it is confusing, so perhaps we need some comments?